src/shogun/distance/EuclideanDistance.cpp

```diff
@@ -66,10 +93,63 @@ float64_t CEuclideanDistance::compute(int32_t idx_a, int32_t idx_b)
 	return CMath::sqrt(result);
 }
 
+void CEuclideanDistance::precompute_rhs_squared_norms()
+{
+	m_rhs_sq_norms=true;
+	SGVector<float64_t>rhs_sq=SGVector<float64_t>(rhs->get_num_vectors());
+	
+	for(int32_t idx_i =0; idx_i<rhs->get_num_vectors(); idx_i++)
+	{
+		SGVector<float64_t> tempvec=((CDenseFeatures<float64_t>*) rhs)->get_feature_vector(idx_i);
+		float64_t temp=linalg::dot(tempvec, tempvec);
+		rhs_sq[idx_i]=temp;
+		((CDenseFeatures<float64_t>*) rhs)->free_feature_vector(tempvec, idx_i);
+	}
+
+	m_rhs_squared_norms=rhs_sq;
+}
+
+void CEuclideanDistance::precompute_lhs_squared_norms()
+{
+	m_lhs_sq_norms=true;
+	SGVector<float64_t>lhs_sq=SGVector<float64_t>(lhs->get_num_vectors());
+
+	for(int32_t idx_i=0; idx_i<lhs->get_num_vectors(); idx_i++)
+	{
+		SGVector<float64_t> tempvec=((CDenseFeatures<float64_t>*) lhs)->get_feature_vector(idx_i);
+		float64_t temp=linalg::dot(tempvec, tempvec);
+		lhs_sq[idx_i]=temp;
+		((CDenseFeatures<float64_t>*) lhs)->free_feature_vector(tempvec, idx_i);
```

Comment:
Can you investigate the `free_feature_vector` mechanism a bit and see whether this call is really needed?